src/scenic/syntax/veneer.py

```diff
@@ -111,1000 +132,1341 @@
 loadingModel = False
 currentSimulation = None
 inInitialScenario = True
-runningScenarios = set()
+runningScenarios = []  # in order, oldest first
 currentBehavior = None
 simulatorFactory = None
 evaluatingGuard = False
+mode2D = False
+_originalConstructibles = (Point, OrientedPoint, Object)
 
 ## APIs used internally by the rest of Scenic
 
 # Scenic compilation
 
 def isActive():
-	"""Are we in the middle of compiling a Scenic module?
-
-	The 'activity' global can be >1 when Scenic modules in turn import other
-	Scenic modules.
-	"""
-	return activity > 0
-
-def activate(paramOverrides={}, modelOverride=None, filename=None, namespace=None):
-	"""Activate the veneer when beginning to compile a Scenic module."""
-	global activity, _globalParameters, lockedParameters, lockedModel, currentScenario
-	if paramOverrides or modelOverride:
-		assert activity == 0
-		_globalParameters.update(paramOverrides)
-		lockedParameters = set(paramOverrides)
-		lockedModel = modelOverride
-
-	activity += 1
-	assert not evaluatingRequirement
-	assert not evaluatingGuard
-	assert currentSimulation is None
-	# placeholder scenario for top-level code
-	newScenario = DynamicScenario._dummy(filename, namespace)
-	scenarioStack.append(newScenario)
-	currentScenario = newScenario
+    """Are we in the middle of compiling a Scenic module?
+
+    The 'activity' global can be >1 when Scenic modules in turn import other
+    Scenic modules.
+    """
+    return activity > 0
+
+def activate(options, namespace=None):
+    """Activate the veneer when beginning to compile a Scenic module."""
+    global activity, _globalParameters, lockedParameters, lockedModel, currentScenario
+    if options.paramOverrides or options.modelOverride:
+        assert activity == 0
+        _globalParameters.update(options.paramOverrides)
+        lockedParameters = set(options.paramOverrides)
+        lockedModel = options.modelOverride
+
+    # If we are in 2D mode, set the global flag and replace all classes
+    # with their 2D compatibility counterparts.
+    if options.mode2D:
+        global mode2D, Point, OrientedPoint, Object
+        mode2D = True
```

Comment:
Let's put an assertion here that either `mode2D` is already true or `activity` is zero; we don't want to somehow compile part of a scenario in 2D mode and another part not in 2D mode.